sunpy/map/mapbase.py

```diff
@@ -1559,66 +1561,66 @@ def draw_limb(self, axes=None, **kwargs):
-    def draw_rectangle(self, bottom_left, width: u.deg, height: u.deg, axes=None, **kwargs):
+    def draw_rectangle(self, bottom_left, *, width: u.deg=None, height: u.deg=None,
+                       axes=None, top_right=None, **kwargs):
         """
         Draw a rectangle defined in world coordinates on the plot.
-
         Parameters
         ----------
-
-        bottom_left : `~astropy.coordinates.SkyCoord` or `~astropy.coordinates.BaseCoordinateFrame`
-            The bottom left corner of the rectangle.
-
-        width : `astropy.units.Quantity`
+        bottom_left : `astropy.units.Quantity` or `~astropy.coordinates.SkyCoord`
+            The bottom_left coordinate of the rectangle. If a `SkyCoord` it can
+            have shape ``(2,)`` and also define ``top_right``. If specifying
+            pixel coordinates it must be given as an `~astropy.units.Quantity`
+            object with units of `~astropy.units.pixel`.
+        top_right : `astropy.units.Quantity` or `~astropy.coordinates.SkyCoord`
+            The top_right coordinate of the rectangle. Can only be omitted if
+            ``bottom_left`` has shape ``(2,)``.
+        width : `astropy.units.Quantity` # passing it as a position argument is deprecated.
             The width of the rectangle.
-
-        height : `astropy.units.Quantity`
+        height : `astropy.units.Quantity` # passing it as a position argument is deprecated.
             The height of the rectangle.
-
         axes : `matplotlib.axes.Axes`
             The axes on which to plot the rectangle, defaults to the current
             axes.
-
         Returns
         -------
-
         rect : `list`
             A list containing the `~matplotlib.patches.Rectangle` object, after
             it has been added to ``axes``.
-
         Notes
         -----
-
         Extra keyword arguments to this function are passed through to the
         `~matplotlib.patches.Rectangle` instance.
-
         """
+        if any(x is None for x in (width, height)):
+            bottom_left, top_right = get_rectangle_coordinates(bottom_left,
+                                                               top_right=top_right,
+                                                               width=width,
+                                                               height=height)
+
+            width = np.abs(top_right.spherical.lon - bottom_left.spherical.lon).to(u.deg)
+            height = np.abs(top_right.spherical.lat - bottom_left.spherical.lat).to(u.deg)
 
         if not axes:
             axes = plt.gca()
-
         if wcsaxes_compat.is_wcsaxes(axes):
             axes_unit = u.deg
         else:
             axes_unit = self.spatial_units[0]
 
         coord = bottom_left.transform_to(self.coordinate_frame)
-        bottom_left = u.Quantity((coord.data.lon, coord.data.lat),
-                                 unit=axes_unit).value
+        bottom_left = u.Quantity((coord.data.lon, coord.data.lat), unit=axes_unit).value
```

Comment:
@ayshih Should this be `.data` or `.spherical`?

Reply:
Hmm.  I want to say `.spherical`, but I suspect here and elsewhere there will be undesirable behavior for heliographic maps and the handling of longitude wraps.  This ought to be tested.

Reply:
Score one for past Albert!  `.spherical` should work fine (at least for SunPy's frames), so change the `.data` to `.spherical`.